examples/pull.rs

```diff
@@ -0,0 +1,186 @@
+) -> Result<(), git2::Error> {
+    let local_tree = repo.find_tree(local.id())?;
+    let remote_tree = repo.find_tree(remote.id())?;
+    let ancestor = repo.find_tree(repo.merge_base(local.id(), remote.id())?)?;
+    let mut idx = repo.merge_trees(&ancestor, &local_tree, &remote_tree, None)?;
+
+    if idx.has_conflicts() {
+        // TODO(jwall): Handle conflicts somehow
+    }
+    let result_tree = repo.find_tree(idx.write_tree_to(repo)?)?;
+    // now create the merge commit
+    let msg = format!("Merge: {} into {}", remote.id(), local.id());
+    let sig = repo.signature()?;
+    let local_commit = repo.find_commit(local.id())?;
+    let remote_commit = repo.find_commit(remote.id())?;
+    // Do our merge commit and set current branch head to that commit.
+    let _merge_commit = repo.commit(
+        Some("HEAD"),
+        &sig,
+        &sig,
+        &msg,
+        &result_tree,
+        &[&local_commit, &remote_commit],
+    )?;
+    // Set working tree to match head.
+    repo.checkout_head(None)?;
+    Ok(())
+}
+
+fn do_merge<'a>(repo: &'a Repository, remote_branch: &str, fetch_commit: git2::AnnotatedCommit<'a>) -> Result<(), git2::Error> {
+        // 1. do a merge analysis
+        let analysis = repo.merge_analysis(&[&fetch_commit])?;
+
+        // 2. Do the appopriate merge
+        if analysis.0.is_fast_forward() {
+            // do a fast forward
+            match repo.find_reference(remote_branch) {
+                Ok(r) => {
+                    let head_commit = repo.reference_to_annotated_commit(&repo.head()?)?;
+                    fast_forward(&mut repo.find_reference(remote_branch)?, &fetch_commit)?;
+                }
+                Err(_) => {
+                    // The branch doesn't exist so just set the reference to the
+                    // commit direcly. Usually this is because you are pulling
+                    // into an empty repository.
+                    let refname = format!("refs/heads/{}", remote_branch);
+                    repo.reference(
+                        &refname,
+                        fetch_commit.id(),
+                        true,
+                        &format!("Setting {} to {}", remote_branch, fetch_commit.id()),
+                    )?;
+                    repo.set_head(&refname)?;
+                }
+            };
+        } else if analysis.0.is_normal() {
+            // do a normal merge
+            let head_commit = repo.reference_to_annotated_commit(&repo.head()?)?;
+            normal_merge(&repo, &head_commit, &fetch_commit)?;
+        } else {
+            println!("Nothing to do...");
+        }
+        Ok(())
+}
+
+fn run(args: &Args) -> Result<(), git2::Error> {
+    let remote_name = args.arg_remote.as_ref().map(|s| &s[..]).unwrap_or("origin");
+    let remote_branch = args.arg_branch.as_ref().map(|s| &s[..]).unwrap_or("master");
+    let repo = Repository::open(".")?;
+    let mut remote = repo.find_remote(&remote_name)?;
```

Comment:
Just realized that this section won't compile. I'll push out a new change later today.